x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java

```diff
@@ -647,48 +625,64 @@ public void setFrequency(TimeValue frequency) {
-                    throw ExceptionsHelper.badRequestException(
-                        Messages.getMessage(Messages.DATAFEED_CONFIG_QUERY_BAD_FORMAT,
-                            id,
-                            exception.getCause().getMessage()),
-                        exception.getCause());
+                QUERY_TRANSFORMER.fromMap(query);
+            } catch(Exception ex) {
+                String msg = Messages.getMessage(Messages.DATAFEED_CONFIG_QUERY_BAD_FORMAT, id);
+
+                if (ex.getCause() instanceof IllegalArgumentException) {
+                    ex = (Exception)ex.getCause();
+                }
+
+                if (lenient) {
+                    logger.warn(msg, ex);
                 } else {
-                    throw ExceptionsHelper.badRequestException(
-                        Messages.getMessage(Messages.DATAFEED_CONFIG_QUERY_BAD_FORMAT, id, exception.getMessage()), exception);
+                    throw ExceptionsHelper.badRequestException(msg, ex);
                 }
             }
         }
 
-        public void setQuery(Map<String, Object> query) {
-            this.query = ExceptionsHelper.requireNonNull(query, QUERY.getPreferredName());
-        }
-
+        // Kept for easier testing
         public void setParsedAggregations(AggregatorFactories.Builder aggregations) {
             try {
                 setAggregations(AGG_TRANSFORMER.toMap(aggregations));
-            } catch (IOException | XContentParseException exception) {
+            } catch (Exception exception) {
                 // Certain thrown exceptions wrap up the real Illegal argument making it hard to determine cause for the user
                 if (exception.getCause() instanceof IllegalArgumentException) {
-                    throw ExceptionsHelper.badRequestException(
-                        Messages.getMessage(Messages.DATAFEED_CONFIG_AGG_BAD_FORMAT,
-                            id,
-                            exception.getCause().getMessage()),
-                        exception.getCause());
-                } else {
-                    throw ExceptionsHelper.badRequestException(
-                        Messages.getMessage(Messages.DATAFEED_CONFIG_AGG_BAD_FORMAT, id, exception.getMessage()), exception);
+                    exception = (Exception)exception.getCause();
                 }
+                throw ExceptionsHelper.badRequestException(
+                    Messages.getMessage(Messages.DATAFEED_CONFIG_AGG_BAD_FORMAT, id), exception);
             }
         }
 
         void setAggregations(Map<String, Object> aggregations) {
+            setAggregations(aggregations, true);
+        }
+
+        void setAggregations(Map<String, Object> aggregations, boolean lenient) {
             this.aggregations = aggregations;
+            try {
+                AGG_TRANSFORMER.fromMap(aggregations);
```

Comment:
Will this throw an exception if an empty map is supplied?  If it does that's good.  If not we should add a check.

Reply:
in data frame transforms I get an "IllegalArgumentException query malformed", this implementation is a little different but I think the same will happen

Reply:
So, for the query parser, it will do this. For an empty agg map, nothing is thrown. I will adjust the behavior so that if the aggs passed are an emptyMap it will throw (if not lenient).